### ISSUE TITLE
lint: transmuting known null pointer to ref 

### ISSUE THREAD

**Issue by llogiq** [Manishearth/rust-clippy#628]
`mem::transmute(std::ptr::null())` – this is undefined behavior.

**Comment by felix91gr** [comment]
I think I can try this one while Oli's busy (I'm atm also working on #3803 under his mentorship). Would you be willing to mentor me here? :)

**Comment by flip1995** [comment]
This is practically just two matches on methods. First you have to find calls to `transmute` and then you have to match the first argument of the call, whether it's a call to `null`. That should get you started. There also may be other methods to create a null pointer like casting `0 as *const _`.

If you got any questions, just ask or open a WIP PR. I'm happy to help you from there.

**Comment by flip1995** [comment]
This is practically just two matches on methods. First you have to find calls to `transmute` and then you have to match the first argument of the call, whether it's a call to `null`. That should get you started. There also may be other methods to create a null pointer like casting `0 as *const _`.

If you got any questions, just ask or open a WIP PR. I'm happy to help you from there.

**Comment by felix91gr** [comment]
Okay I'm now making a first UI test for this. It should show all the cases that we're looking for in this lint.

Which cases are missing?

```rust

fn transmute_0() {
    let zero : *const i32 = 0 as *const_;
    let zero_ts : &i32 = std::mem::transmute(zero);
}

fn transmute_null() {
    let nil : *const i32 = std::ptr::null();
    let nil_ts : &i32 = std::mem::transmute(nil);
}

fn main() {
    transmute_0();
    transmute_null();
}

```

**Comment by flip1995** [comment]
Maybe you can also add:
```rust
const ZPTR: *const usize = 0 as *const _;
... std::mem::transmute(ZPTR); ...

... std::mem::transmute(0 as *const _); ...
```

Maybe also some different types, but that shouldn't be an issue on getting this lint working.

**Comment by felix91gr** [comment]
Thank you!

This should use a `LateLintPass`, right? Since this is more than just a syntactic analysis.

**Comment by felix91gr** [comment]
Working on this at #3848

**Comment by lzutao** [comment]
Could we close this now?

**Comment by felix91gr** [comment]
Maybe? I'm not sure, @lzutao. The thing is, my PR implemented this but only the trivial bits. There are many cases where one would see that a null pointer was transmuted, but we can't lint those yet because `miri` isn't yet as complete as we needed.

The idea was that at some point we could do simple constant propagation over a stretch of code using `miri`, and then check if `transmute` was being called with any propagated `null` pointers. This is very hard and not necessary to do by hand, but doing it with `miri` would be trivial. That's why we chose to wait on it before completing this feature :3

**Comment by felix91gr** [comment]
Is there anyone working on this at the moment? If I understand correctly, the constant propagation machinery should now be able to propagate the `null` pointers. Am I right, @oli-obk?

If so, and if no one is working on this, then I'd like to get back on this so we can close this issue :D

**Comment by flip1995** [comment]
Nope no one is working on this right now.

**Comment by felix91gr** [comment]
Awesome. If Oliver confirms to me that this can be completed now, I'll tackle it :)

**Comment by oli-obk** [comment]
Unfortunately not yet, but I don't know why. If you click on the three dots in the top left corner and select MIR https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=907e81a3947910a850f01271a054209e you'll see 

```rust
bb0: {
        StorageLive(_1);                 // bb0[0]: scope 1 at src/main.rs:3:13: 3:14
        _1 = const 0usize;               // bb0[1]: scope 1 at src/main.rs:3:17: 3:18
                                         // ty::Const
                                         // + ty: usize
                                         // + val: Value(Scalar(0x0000000000000000))
                                         // mir::Constant
                                         // + span: src/main.rs:3:17: 3:18
                                         // + literal: Const { ty: usize, val: Value(Scalar(0x0000000000000000)) }
        StorageLive(_2);                 // bb0[2]: scope 2 at src/main.rs:4:9: 4:46
        StorageLive(_3);                 // bb0[3]: scope 2 at src/main.rs:4:44: 4:45
        _3 = _1;                         // bb0[4]: scope 2 at src/main.rs:4:44: 4:45
        _2 = const std::intrinsics::transmute::<usize, &i32>(move _3) -> bb1; // bb0[5]: scope 2 at src/main.rs:4:9: 4:46
                                         // ty::Const
                                         // + ty: unsafe extern "rust-intrinsic" fn(usize) -> &i32 {std::intrinsics::transmute::<usize, &i32>}
                                         // + val: Value(Scalar(<ZST>))
                                         // mir::Constant
                                         // + span: src/main.rs:4:9: 4:43
                                         // + user_ty: UserType(0)
                                         // + literal: Const { ty: unsafe extern "rust-intrinsic" fn(usize) -> &i32 {std::intrinsics::transmute::<usize, &i32>}, val: Value(Scalar(<ZST>)) }
    }
```

which (ignorign a lot of the comments and `StorageLive` identifiers, doesn't really have any reason why the `0usize isn't propagated into the function call

```
_2 = const std::intrinsics::transmute::<usize, &i32>(move _3) -> bb1;
```

I'd expect the above to be

```
_2 = const std::intrinsics::transmute::<usize, &i32>(0usize) -> bb1;
```

which would then be easy to detect. This needs changes in rustc, but if you're interested, I can mentor you for that. I believe it's not very hard to do, but you'll have to learn quite a bit about MIR.

**Comment by felix91gr** [comment]
Excellent. I accept your mentorship, @oli-obk :rainbow: 

Let's do this.

Also: hey, this will serve as a future test for const propagation :D

**Comment by RalfJung** [comment]
FWIW, this seems very similar to the existing "invalid_value" lint in rustc (not sure if Clippy also has such a lint), which already detects `transmute(0usize)` to a reference.

**Comment by felix91gr** [comment]
Oh, that's cool! I think this is more general maybe? Although with dataflow-aware constprop, the `invalid_value` lint might become equivalent to this, depending on how it's constructed :3

**Comment by RalfJung** [comment]
`invalid_value` is a HIR lint so I doubt constprop will change anything there. ;)

However, [`is_zero`](https://github.com/rust-lang/rust/blob/90ccf4f5adfb2562fc95c996b97faac7775a34bb/compiler/rustc_lint/src/builtin.rs#L2355) could be extended to recognize `ptr::null`.

**Comment by Jarcho** [comment]
Closing this since the lint does exist.

**Comment by felix91gr** [comment]
@Jarcho I have a question about lint policy.

The lint is known to be incomplete, as I noted in the [documentation](https://rust-lang.github.io/rust-clippy/master/index.html#transmuting_null) eons ago when I tried my hand at implementing it.

Regarding that, the lint is always going to be incomplete because of the Halting Problem. That's fine. But its current version is very shallow.

The ideal form of it is something that uses some kind of value-range analysis or const folding in order to deduce a larger portion of the pointers fed to `transmute` which are **always** going to be null. I tried adding that kind of machinery & I couldn't pull it off, however it's very reasonable to expect that kind of thing to eventually be added to the compiler because it's such a useful analysis framework in general. I believe this is why I didn't close this issue. I left it open as a pointer to the work that was yet to be done.

So coming back to policy: knowing this, should...
1. This issue remain opened?
2. This issue be closed, and another issue detailing this future work be opened?
3. This issue be closed, and another issue with future work for lints be opened (or edited) to document this?
4. This issue be closed, and the documentation in the lint be left as the canonical pointer to "future work"?

Does that make sense? Sorry I'm opening this issue back up. I basically want to know what's the policy for, let's call them, "partially fulfilled lint implementations": impls that we know can be substantially better, that are however held back from achieving that at the moment, for one reason or another.

---

Also, good god, **thanks** for doing triage on so many opened issues. I wanted to help in this regard a few months back, since Clippy has such a large backlog of opened issues. However, my lack of deep holistic knowledge of Clippy became quite the obvious impediment. You're way better suited for this job than I am 😄

**Comment by Jarcho** [comment]
Short answer: Make a new issue.

Much longer answer:

If a lint exists that reasonably fulfills a lint request then there shouldn't be an issue open for that. Allowing the fact that the lint could be extended to keep an issue open would result in a fair number of nigh unclosable issues. This is due to a large number of lints have a vague enough description of what they detect that we will never detect the full set.

For an issue to be useful it has to either be actionable (as in there is a clear action to take), or it has to act as a reminder for something that needs to known (not a fan of using github's issues for this, but people do it). That a lint could, in some way, be improved or extended doesn't fit either of these. It's just a statement of the assumed default state. Even worse for keeping lint requests open is that they'll be actively misleading. They make it seem like something could be added even though it already exists, the discussion will start about adding such a thing, and somewhere buried will be the part that's still relevant. You can't really change the title or first post to fix part of this without making digging through history significantly harder either.

For making a new issue about extending a lint as long as there is a clear action to take (even if nobody gets around to doing it) then it's useful. A clear description of how a lint could be better meets this.
